File: aoc/2024/aoc202015.hs
module AOC202015 where

import Control.Monad.Primitive(PrimState)
import Control.Monad.ST(ST,runST)
import Data.Vector.Unboxed.Mutable(MVector,write)
import qualified Data.Vector.Unboxed.Mutable as Vector

import AOC

aoc = AOC {
    day="../../2020/input/15",
    -- part 2 code is too slow for the tests
    aocTests=[
        AOCTest {
            testData="0,3,6",
            testResult=Just "436",
            testResult2=Nothing --Just "175594"
            },
        AOCTest {
            testData="1,3,2",
            testResult=Just "1",
            testResult2=Nothing --Just "2578"
            },
        AOCTest {
            testData="2,1,3",
            testResult=Just "10",
            testResult2=Nothing --Just "3544142"
            },
        AOCTest {
            testData="1,2,3",
            testResult=Just "27",
            testResult2=Nothing --Just "261214"
            },
        AOCTest {
            testData="2,3,1",
            testResult=Just "78",
            testResult2=Nothing --Just "6895259"
            },
        AOCTest {
            testData="3,2,1",
            testResult=Just "438",
            testResult2=Nothing --Just "18"
            },
        AOCTest {
            testData="3,1,2",
            testResult=Just "1836",
            testResult2=Nothing --Just "362"
            }
        ],
    aocCode=Code {
        codeParse=parseInts,
        codeParse2=parseInts,
        codeTest=result 2020,
        codeTest2=result 30000000,
        codeResult=result 2020,
        codeResult2=result 30000000
        }
    }

speak :: Int -> Int -> Int -> MVector (PrimState (ST s)) Int -> ST s Int
speak n turn number history
  | turn >= n = return number
  | otherwise = do
      lastSpoken <- Vector.read history number
      write history number turn
      if lastSpoken == 0
        then speak n (turn+1) 0 history
        else speak n (turn+1) (turn - lastSpoken) history

starting :: Int -> [Int] -> ST s Int
starting n numbers = do
    history <- Vector.replicate n 0
    sequence_ [write history num i | (i,num) <- zip [1..] (init numbers)]
    speak n (length numbers) (last numbers) history

result n numbers = runST $ starting n numbers

File: aoc/2025/aoc202015.go
package main

func init() {
	Register(&aoc202015{
		AOC: AOC{
			Day:           15,
			InputFilename: "../2020/input/15.txt",
			Tests: []Test{
				Test{
					Input: `0,3,6
`,
					Part1: "436",
					Part2: "", //"175594",
				},
				Test{
					Input: `1,3,2
`,
					Part1: "1",
					Part2: "", //"2578",
				},
				Test{
					Input: `2,1,3
`,
					Part1: "10",
					Part2: "", //"3544142",
				},
				Test{
					Input: `1,2,3
`,
					Part1: "27",
					Part2: "", //"261214",
				},
				Test{
					Input: `2,3,1
`,
					Part1: "78",
					Part2: "", //"6895259",
				},
				Test{
					Input: `3,2,1
`,
					Part1: "438",
					Part2: "", //"18",
				},
				Test{
					Input: `3,1,2
`,
					Part1: "1836",
					Part2: "", //"362",
				},
			},
		},
	})
}

type aoc202015 struct {
	AOC
}

func (aoc *aoc202015) result(limit int, input *Input) string {
	turn := 0
	history := make([]int, limit)
	last := 0
	for n := range input.IntSeq() {
		if turn > 0 {
			history[last] = turn
		}
		turn++
		last = n
	}
	for turn < limit {
		n := 0
		if history[last] > 0 {
			n = turn - history[last]
		}
		history[last] = turn
		turn++
		last = n
	}
	return IntResult(last)
}

func (aoc *aoc202015) Part1(input *Input) string {
	return aoc.result(2020, input)
}

func (aoc *aoc202015) Part2(input *Input) string {
	return aoc.result(30000000, input)
}
